components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java

```diff
@@ -355,36 +357,58 @@ public static String getEndpointURIPath(String endpoint, boolean addProxyContext
-        String hostName = ServerConfiguration.getInstance().getFirstProperty(IdentityCoreConstants.HOST_NAME);
 
+        return getServerURL(urlContext, addProxyContextPath, addWebContextRoot, false);
+    }
+
+    /**
+     * This method is used to return a URL with a proxy context path, a web context root and the tenant domain (If
+     * required) when provided with a URL.
+     * @param url URL.
+     * @param addProxyContextPath add proxy context path to the URL.
+     * @param addWebContextRoot add web context path to the URL.
+     * @return
+     */
+    public static String resolveURL(String url, boolean addProxyContextPath, boolean addWebContextRoot) {
+
+        URLResolverService urlResolverService = IdentityCoreServiceComponent.getURLResolverService();
         try {
-            if (hostName == null) {
-                hostName = NetworkUtils.getLocalHostname();
-            }
-        } catch (SocketException e) {
-            throw IdentityRuntimeException.error("Error while trying to read hostname.", e);
+            return urlResolverService.resolveUrl(url, addProxyContextPath, addWebContextRoot, null);
+        } catch (URLResolverException e) {
+            throw IdentityRuntimeException.error("Error while resolving URL: " + url, e);
         }
+    }
 
-        String mgtTransport = CarbonUtils.getManagementTransport();
-        AxisConfiguration axisConfiguration = IdentityCoreServiceComponent.getConfigurationContextService().
-                getServerConfigContext().getAxisConfiguration();
-        int mgtTransportPort = CarbonUtils.getTransportProxyPort(axisConfiguration, mgtTransport);
-        if (mgtTransportPort <= 0) {
-            mgtTransportPort = CarbonUtils.getTransportPort(axisConfiguration, mgtTransport);
-        }
-        if (hostName.endsWith("/")) {
-            hostName = hostName.substring(0, hostName.length() - 1);
-        }
-        StringBuilder serverUrl = new StringBuilder(mgtTransport).append("://").append(hostName.toLowerCase());
-        // If it's well known HTTPS port, skip adding port
-        if (mgtTransportPort != IdentityCoreConstants.DEFAULT_HTTPS_PORT) {
-            serverUrl.append(":").append(mgtTransportPort);
-        }
+    /**
+     * This util method is used to construct a complete URL out of the given URL context.
+     * @param urlContext URL context.
+     * @param addProxyContextPath add proxy context path to the URL.
+     * @param addWebContextRoot add web context path to the URL.
+     * @param addTenantQueryParamInLegacyMode add tenant domain as a query parameter in legacy mode.
+     * @return complete URL for the given URL context.
+     * @throws IdentityRuntimeException if error occurred while constructing the URL.
+     */
+    public static String getServerURL(String urlContext, boolean addProxyContextPath,
+                                                     boolean addWebContextRoot, boolean addTenantQueryParamInLegacyMode)
```

Comment:
Let's use the same param name through out. 'addTenantQueryParamInLegacyMode' is more readable and meaningful than the name used before